File: ios-app/ICAV Time Tracker/ICAV Time Tracker/TimeEntry.swift
//
//  TimeEntry.swift
//  ICAV Time Tracker
//
//  Created by Jason Baker on 6/23/25.
//

import Foundation

struct TimeEntry: Identifiable, Codable {
    var id: UUID
    let userId: String
    let technicianName: String
    let customerName: String
    var clockInTime: Date?
    var clockOutTime: Date?
    var lunchStartTime: Date?
    var lunchEndTime: Date?
    var driveStartTime: Date?
    var driveEndTime: Date?
    
    // Sync tracking
    var serverId: String? // ID from the server after successful sync
    var isSynced: Bool = false // Whether this entry has been synced to server
    var needsSync: Bool = false // Whether this entry has local changes that need syncing
    var markedForDeletion: Bool = false // Whether this entry should be deleted from server
    var lastModified: Date = Date() // When this entry was last modified locally
    
    // Custom initializer to generate UUID
    init(userId: String, technicianName: String, customerName: String, clockInTime: Date, clockOutTime: Date? = nil, lunchStartTime: Date? = nil, lunchEndTime: Date? = nil, driveStartTime: Date? = nil, driveEndTime: Date? = nil) {
        self.id = UUID()
        self.userId = userId
        self.technicianName = technicianName
        self.customerName = customerName
        self.clockInTime = clockInTime
        self.clockOutTime = clockOutTime
        self.lunchStartTime = lunchStartTime
        self.lunchEndTime = lunchEndTime
        self.driveStartTime = driveStartTime
        self.driveEndTime = driveEndTime
    }
    
    // Initializer for driving entries (without clockInTime)
    init(userId: String, technicianName: String, customerName: String, driveStartTime: Date) {
        self.id = UUID()
        self.userId = userId
        self.technicianName = technicianName
        self.customerName = customerName
        self.clockInTime = nil
        self.clockOutTime = nil
        self.lunchStartTime = nil
        self.lunchEndTime = nil
        self.driveStartTime = driveStartTime
        self.driveEndTime = nil
    }
    
    // New initializer for a job with no clock-in or drive time
    init(userId: String, technicianName: String, customerName: String) {
        self.id = UUID()
        self.userId = userId
        self.technicianName = technicianName
        self.customerName = customerName
        self.clockInTime = nil
        self.clockOutTime = nil
        self.lunchStartTime = nil
        self.lunchEndTime = nil
        self.driveStartTime = nil
        self.driveEndTime = nil
        self.serverId = nil
        self.isSynced = false
        self.needsSync = false
        self.lastModified = Date()
    }
    
    var isActive: Bool {
        return clockOutTime == nil && (clockInTime != nil || driveStartTime != nil)
    }
    
    var isOnLunch: Bool {
        return lunchStartTime != nil && lunchEndTime == nil
    }
    
    var isDriving: Bool {
        return driveStartTime != nil && driveEndTime == nil
    }
    
    var duration: TimeInterval? {
        guard let clockOutTime = clockOutTime else { return nil }
        return clockOutTime.timeIntervalSince(clockInTime ?? Date.distantPast)
    }
    
    var formattedDuration: String? {
        guard let duration = duration else { return nil }
        let hours = Int(duration) / 3600
        let minutes = Int(duration) % 3600 / 60
        return String(format: "%02d:%02d", hours, minutes)
    }
    
    var lunchDuration: TimeInterval? {
        guard let lunchStart = lunchStartTime, let lunchEnd = lunchEndTime else { return nil }
        return lunchEnd.timeIntervalSince(lunchStart)
    }
    
    var formattedLunchDuration: String? {
        guard let duration = lunchDuration else { return nil }
        let hours = Int(duration) / 3600
        let minutes = Int(duration) % 3600 / 60
        return String(format: "%02d:%02d", hours, minutes)
    }
    
    var driveDuration: TimeInterval? {
        guard let driveStart = driveStartTime, let driveEnd = driveEndTime else { return nil }
        return driveEnd.timeIntervalSince(driveStart)
    }
    
    var formattedDriveDuration: String? {
        guard let duration = driveDuration else { return nil }
        let hours = Int(duration) / 3600
        let minutes = Int(duration) % 3600 / 60
        return String(format: "%02d:%02d", hours, minutes)
    }
    
    // Helper methods for sync management
    mutating func markForSync() {
        needsSync = true
        lastModified = Date()
    }
    
    mutating func markAsSynced(serverId: String) {
        self.serverId = serverId
        isSynced = true
        needsSync = false
    }
    
    mutating func markForDeletion() {
        markedForDeletion = true
        needsSync = true
        lastModified = Date()
    }
    
    var syncStatus: String {
        if isSynced {
            return "✅ Synced"
        } else if needsSync {
            return "📤 Pending sync"
        } else {
            return "📱 Local only"
        }
    }
}

enum ClockStatus {
    case clockedOut
    case driving
    case clockedIn(TimeEntry)
    case onLunch(TimeEntry)
} 

File: android-app/Android App/app/src/main/java/com/example/icavtimetracker/data/TimeEntry.kt
package com.example.icavtimetracker.data

import com.google.gson.annotations.SerializedName
import java.util.Date
import java.util.UUID

data class TimeEntry(
    val id: String = UUID.randomUUID().toString(),
    @SerializedName("userId")
    val userId: String,
    @SerializedName("technicianName")
    val technicianName: String,
    @SerializedName("customerName")
    val customerName: String,
    @SerializedName("clockInTime")
    val clockInTime: Date? = null,
    @SerializedName("clockOutTime")
    var clockOutTime: Date? = null,
    @SerializedName("lunchStartTime")
    var lunchStartTime: Date? = null,
    @SerializedName("lunchEndTime")
    var lunchEndTime: Date? = null,
    @SerializedName("driveStartTime")
    var driveStartTime: Date? = null,
    @SerializedName("driveEndTime")
    var driveEndTime: Date? = null,
    
    // Sync tracking
    @SerializedName("serverId")
    var serverId: String? = null,
    var isSynced: Boolean = false,
    var needsSync: Boolean = false,
    var lastModified: Date = Date()
) {
    val isActive: Boolean
        get() = clockOutTime == null
    
    val isOnLunch: Boolean
        get() = lunchStartTime != null && lunchEndTime == null
    
    val isDriving: Boolean
        get() = driveStartTime != null && driveEndTime == null
    
    val duration: Long?
        get() = clockOutTime?.let { clockOut ->
            clockInTime?.let { clockIn ->
                clockOut.time - clockIn.time
            }
        }
    
    val formattedDuration: String?
        get() = duration?.let { durationMs ->
            val hours = (durationMs / (1000 * 60 * 60)).toInt()
            val minutes = ((durationMs % (1000 * 60 * 60)) / (1000 * 60)).toInt()
            String.format("%02d:%02d", hours, minutes)
        }
    
    val lunchDuration: Long?
        get() = if (lunchStartTime != null && lunchEndTime != null) {
            lunchEndTime!!.time - lunchStartTime!!.time
        } else null
    
    val formattedLunchDuration: String?
        get() = lunchDuration?.let { durationMs ->
            val hours = (durationMs / (1000 * 60 * 60)).toInt()
            val minutes = ((durationMs % (1000 * 60 * 60)) / (1000 * 60)).toInt()
            String.format("%02d:%02d", hours, minutes)
        }
    
    val driveDuration: Long?
        get() = if (driveStartTime != null && driveEndTime != null) {
            driveEndTime!!.time - driveStartTime!!.time
        } else null
    
    val formattedDriveDuration: String?
        get() = driveDuration?.let { durationMs ->
            val hours = (durationMs / (1000 * 60 * 60)).toInt()
            val minutes = ((durationMs % (1000 * 60 * 60)) / (1000 * 60)).toInt()
            String.format("%02d:%02d", hours, minutes)
        }
    
    fun markForSync() {
        needsSync = true
        lastModified = Date()
    }
    
    fun markAsSynced(serverId: String) {
        this.serverId = serverId
        isSynced = true
        needsSync = false
    }
    
    val syncStatus: String
        get() = when {
            isSynced -> "✅ Synced"
            needsSync -> "📤 Pending sync"
            else -> "📱 Local only"
        }
}

enum class ClockStatus {
    CLOCKED_OUT,
    DRIVING,
    CLOCKED_IN,
    ON_LUNCH
} 